// @remote-dom/core/build/cjs/constants.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const NODE_TYPE_ELEMENT = 1;
const NODE_TYPE_TEXT = 3;
const NODE_TYPE_COMMENT = 8;
const NODE_TYPE_ROOT = 9;
const MUTATION_TYPE_INSERT_CHILD = 0;
const MUTATION_TYPE_REMOVE_CHILD = 1;
const MUTATION_TYPE_UPDATE_TEXT = 2;
const MUTATION_TYPE_UPDATE_PROPERTY = 3;
const UPDATE_PROPERTY_TYPE_PROPERTY = 1;
const UPDATE_PROPERTY_TYPE_ATTRIBUTE = 2;
const UPDATE_PROPERTY_TYPE_EVENT_LISTENER = 3;
const ROOT_ID = '~';

exports.MUTATION_TYPE_INSERT_CHILD = MUTATION_TYPE_INSERT_CHILD;
exports.MUTATION_TYPE_REMOVE_CHILD = MUTATION_TYPE_REMOVE_CHILD;
exports.MUTATION_TYPE_UPDATE_PROPERTY = MUTATION_TYPE_UPDATE_PROPERTY;
exports.MUTATION_TYPE_UPDATE_TEXT = MUTATION_TYPE_UPDATE_TEXT;
exports.NODE_TYPE_COMMENT = NODE_TYPE_COMMENT;
exports.NODE_TYPE_ELEMENT = NODE_TYPE_ELEMENT;
exports.NODE_TYPE_ROOT = NODE_TYPE_ROOT;
exports.NODE_TYPE_TEXT = NODE_TYPE_TEXT;
exports.ROOT_ID = ROOT_ID;
exports.UPDATE_PROPERTY_TYPE_ATTRIBUTE = UPDATE_PROPERTY_TYPE_ATTRIBUTE;
exports.UPDATE_PROPERTY_TYPE_EVENT_LISTENER = UPDATE_PROPERTY_TYPE_EVENT_LISTENER;
exports.UPDATE_PROPERTY_TYPE_PROPERTY = UPDATE_PROPERTY_TYPE_PROPERTY;
