// @remote-dom/core/build/cjs/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const connection = require('./connection.cjs');
const constants = require('./constants.cjs');



exports.createRemoteConnection = connection.createRemoteConnection;
exports.MUTATION_TYPE_INSERT_CHILD = constants.MUTATION_TYPE_INSERT_CHILD;
exports.MUTATION_TYPE_REMOVE_CHILD = constants.MUTATION_TYPE_REMOVE_CHILD;
exports.MUTATION_TYPE_UPDATE_PROPERTY = constants.MUTATION_TYPE_UPDATE_PROPERTY;
exports.MUTATION_TYPE_UPDATE_TEXT = constants.MUTATION_TYPE_UPDATE_TEXT;
exports.NODE_TYPE_COMMENT = constants.NODE_TYPE_COMMENT;
exports.NODE_TYPE_ELEMENT = constants.NODE_TYPE_ELEMENT;
exports.NODE_TYPE_ROOT = constants.NODE_TYPE_ROOT;
exports.NODE_TYPE_TEXT = constants.NODE_TYPE_TEXT;
exports.ROOT_ID = constants.ROOT_ID;
exports.UPDATE_PROPERTY_TYPE_ATTRIBUTE = constants.UPDATE_PROPERTY_TYPE_ATTRIBUTE;
exports.UPDATE_PROPERTY_TYPE_EVENT_LISTENER = constants.UPDATE_PROPERTY_TYPE_EVENT_LISTENER;
exports.UPDATE_PROPERTY_TYPE_PROPERTY = constants.UPDATE_PROPERTY_TYPE_PROPERTY;
